### PR TITLE
Add MultiHeadedSimilarity

### PR DESCRIPTION
**Pull request by matt-gardner**
This is a wrapper around any similarity function that projects and splits each tensor into multiple heads before computing the similarity.  It's the first step to getting a multi-headed self-attention encoder; the rest of that should be pretty easy.

I also moved `similarity_function.py` to be inside the `similarity_functions` module.

### CODE REVIEW

File: allennlp/modules/similarity_functions/multiheaded.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        torch.nn.init.xavier_uniform(self._tensor_2_projection)
+
+    @overrides
+    def forward(self, tensor_1: torch.Tensor, tensor_2: torch.Tensor) -> torch.Tensor:
+        projected_tensor_1 = torch.matmul(tensor_1, self._tensor_1_projection)
+        projected_tensor_2 = torch.matmul(tensor_2, self._tensor_2_projection)
+
+        # split_tensor_{1,2} are both tuples of length num_heads, where each element in the tuple
+        # is a tensor of shape (..., projected_dim / num_heads).
+        split_size = projected_tensor_1.size(-1) // self._num_heads
```
Comment (DeNeutoy) [inline allennlp/modules/similarity_functions/multiheaded.py]: You don't need this split and stack - you can just do 
```
all_but_final_dim = projected_tensor_x.size()[:-1]
projected_tensor_x.resize(*all_but_final_dim, self._num_heads, -1)
```

File: allennlp/modules/similarity_functions/multiheaded.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        tensor_1_projected_dim = tensor_1_projected_dim or tensor_1_dim
+        tensor_2_dim = tensor_2_dim or tensor_1_dim
+        tensor_2_projected_dim = tensor_2_projected_dim or tensor_2_dim
+        if tensor_1_projected_dim % num_heads != 0:
+            raise ConfigurationError("Projected dimension not divisible by number of heads: %d, %d"
+                                     % (tensor_1_projected_dim, num_heads))
+        if tensor_2_projected_dim % num_heads != 0:
+            raise ConfigurationError("Projected dimension not divisible by number of heads: %d, %d"
+                                     % (tensor_2_projected_dim, num_heads))
+        self._tensor_1_projection = Parameter(torch.Tensor(tensor_1_dim, tensor_1_projected_dim))
```
Comment (DeNeutoy) [inline allennlp/modules/similarity_functions/multiheaded.py]: Is there a reason you aren't using `torch.nn.modules.Linear` for these?

File: allennlp/modules/similarity_functions/multiheaded.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+        # We then stack the tensors, inserting a new dimension in the _second-to-last_ position,
+        # giving us a tensor of shape (..., num_heads, projected_dim / num_heads).
+        stacked_tensor_1 = torch.stack(split_tensor_1, dim=-2)
+        stacked_tensor_2 = torch.stack(split_tensor_2, dim=-2)
+
+        # And then we pass this off to our internal similarity function.  Because the similarity
+        # functions don't care what dimension their input has, and only look at the last dimension,
+        # we don't need to do anything special here.  It will just compute similarity on the
+        # (..., num_heads, *projected_dim / num_heads*) dimension, returning a tensor of shape
```
Comment (DeNeutoy) [inline allennlp/modules/similarity_functions/multiheaded.py]: Are you just trying to say that the similarity is computed along the projection dimension for each head? This `*`notation is a little confusing.

File: allennlp/modules/similarity_functions/multiheaded.py (hunk 2 above)
Comment (DeNeutoy) [reply allennlp/modules/similarity_functions/multiheaded.py]: Nevermind, the rest of the similarity functions also do this.

File: tests/modules/similarity_functions/multiheaded_test.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+                                           tensor_2_dim=6, tensor_2_projected_dim=12)
+        assert list(similarity._tensor_1_projection.size()) == [9, 6]
+        assert list(similarity._tensor_2_projection.size()) == [6, 12]
+        with pytest.raises(ConfigurationError):
+            similarity = MultiHeadedSimilarity(num_heads=3, tensor_1_dim=10)
+        with pytest.raises(ConfigurationError):
+            params = Params({'num_heads': 3, 'tensor_1_dim': 9, 'tensor_2_dim': 10})
+            MultiHeadedSimilarity.from_params(params)
+
+    def test_forward(self):
```
Comment (DeNeutoy) [inline tests/modules/similarity_functions/multiheaded_test.py]: A test for the 3D case might be nice, given it's the main use case.

File: allennlp/modules/similarity_functions/multiheaded.py (hunk 1 above)
Comment (matt-gardner) [reply allennlp/modules/similarity_functions/multiheaded.py]: Right, good point, thanks.

File: allennlp/modules/similarity_functions/multiheaded.py (hunk 3 above)
Comment (matt-gardner) [reply allennlp/modules/similarity_functions/multiheaded.py]: Done.

File: tests/modules/similarity_functions/multiheaded_test.py (hunk 4 above)
Comment (matt-gardner) [reply tests/modules/similarity_functions/multiheaded_test.py]: The similarity function is totally agnostic to how many dimensions there are.  I added a couple of empty dimensions to the test, but I'm not sure it really adds anything.

**Pull request merged**